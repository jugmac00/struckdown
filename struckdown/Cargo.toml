[package]
name = "struckdown"
version = "0.1.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
edition = "2018"
license = "Apache-2.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
default = ["external-processor"]
external-processor = ["tokio", "subprocess"]

[dependencies]
pulldown-cmark = "0.8.0"
serde = { version = "1.0.118", features = ["derive"] }
regex = "1.4.2"
lazy_static = "1.4.0"
v_htmlescape = "0.11.0"
itertools = "0.9.0"
serde_yaml = "0.8.14"
slug = "0.1.4"
serde_json = { version = "1.0.60", features = ["preserve_order"] }
subprocess = { version = "0.2.6", optional = true }
tokio = { version = "0.3.6", features = ["rt", "process", "macros", "io-util"], optional=true }

[dev-dependencies]
insta = { version = "1.3.0", features = ["glob"] }
