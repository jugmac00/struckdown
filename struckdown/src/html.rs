//! Implements an HTML renderer.
use std::borrow::Cow;
use std::collections::HashMap;
use std::io::{self, Write};

use serde::{Deserialize, Serialize};
use v_htmlescape::escape;

use crate::event::{
    Alignment, AnnotatedEvent, Attrs, CheckboxEvent, CodeBlockEvent, DirectiveEvent, EndTagEvent,
    ErrorEvent, Event, FootnoteReferenceEvent, ImageEvent, InlineCodeEvent, InterpretedTextEvent,
    RawHtmlEvent, StartTagEvent, Str, Tag, TextEvent,
};

/// Customizes the HTML rendering.
#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(default)]
pub struct HtmlRendererOptions {
    /// When enabled `_foo_` renders into underlines.
    pub render_underlines: bool,
    /// The class to emit for footnote references.
    pub footnote_reference_class: String,
    /// The class to emit for footnote definitions.
    pub footnote_definition_class: String,
    /// The initial level for headlines
    pub initial_headline_level: usize,
}

impl Default for HtmlRendererOptions {
    fn default() -> Self {
        Self::new()
    }
}

impl HtmlRendererOptions {
    /// Creates the default configuraiton for the renderer options.
    pub fn new() -> HtmlRendererOptions {
        HtmlRendererOptions {
            render_underlines: false,
            footnote_reference_class: "footnote-reference".into(),
            footnote_definition_class: "footnote-definition".into(),
            initial_headline_level: 1,
        }
    }
}

/// Object capable of rendering events to HTML.
pub struct HtmlRenderer<'data, 'options, F> {
    out: F,
    footnotes: HashMap<Str<'data>, usize>,
    options: &'options HtmlRendererOptions,
}

impl<'data, 'options, F: Write> HtmlRenderer<'data, 'options, F> {
    /// Creates a new renderer that writes into a writer.
    pub fn new(out: F, options: &'options HtmlRendererOptions) -> HtmlRenderer<'data, 'options, F> {
        HtmlRenderer {
            out,
            footnotes: HashMap::new(),
            options,
        }
    }

    /// Consumes the writer and returns the inner file.
    pub fn into_writer(self) -> F {
        self.out
    }

    fn is_block_tag(&self, tag: Tag) -> bool {
        match tag {
            Tag::Paragraph => true,
            Tag::Heading1 => true,
            Tag::Heading2 => true,
            Tag::Heading3 => true,
            Tag::Heading4 => true,
            Tag::Heading5 => true,
            Tag::Heading6 => true,
            Tag::BlockQuote => true,
            Tag::OrderedList => true,
            Tag::UnorderedList => true,
            Tag::ListItem => true,
            Tag::FootnoteDefinition => true,
            Tag::Table => true,
            Tag::TableHead => true,
            Tag::TableRow => true,
            Tag::TableCell => true,
            Tag::Emphasis => false,
            Tag::EmphasisAlt => false,
            Tag::Strong => false,
            Tag::Strikethrough => false,
            Tag::Link => false,
            Tag::TableHeader => true,
            Tag::TableBody => true,
            Tag::Container => true,
            Tag::Span => false,
        }
    }

    fn tag_to_html_tag(&self, tag: Tag) -> &'static str {
        match tag {
            Tag::Paragraph => "p",
            Tag::Heading1
            | Tag::Heading2
            | Tag::Heading3
            | Tag::Heading4
            | Tag::Heading5
            | Tag::Heading6 => {
                let level = tag.header_level().unwrap();
                let target_level = (self.options.initial_headline_level.saturating_sub(1)) + level;
                match target_level {
                    1 => "h1",
                    2 => "h2",
                    3 => "h3",
                    4 => "h4",
                    5 => "h5",
                    _ => "h6",
                }
            }
            Tag::BlockQuote => "blockquote",
            Tag::OrderedList => "ol",
            Tag::UnorderedList => "ul",
            Tag::ListItem => "li",
            Tag::FootnoteDefinition => "div",
            Tag::Table => "table",
            Tag::TableHeader => "thead",
            Tag::TableBody => "tbody",
            Tag::TableRow => "tr",
            Tag::TableHead => "th",
            Tag::TableCell => "td",
            Tag::Emphasis => "em",
            Tag::EmphasisAlt => {
                if self.options.render_underlines {
                    "u"
                } else {
                    "em"
                }
            }
            Tag::Strong => "strong",
            Tag::Strikethrough => "ss",
            Tag::Link => "a",
            Tag::Container => "div",
            Tag::Span => "span",
        }
    }

    fn start_tag(&mut self, tag: Tag, attrs: &Attrs) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);
        write!(self.out, "<{}", html_tag)?;

        match attrs.start {
            None | Some(1) => {}
            Some(start) => {
                write!(self.out, " start={}", start)?;
            }
        }

        let mut combined_style = String::new();
        if let Some(ref id) = attrs.id {
            write!(self.out, " id=\"{}\"", escape(id.as_str()))?;
        }
        if let Some(ref title) = attrs.title {
            write!(self.out, " title=\"{}\"", escape(title.as_str()))?;
        }
        if let Some(ref target) = attrs.target {
            write!(self.out, " href=\"{}\"", escape(target.as_str()))?;
        }

        combined_style.push_str(match attrs.alignment {
            Alignment::None => "",
            Alignment::Left => "text-align: left",
            Alignment::Center => "text-align: center",
            Alignment::Right => "text-align: right",
        });

        let mut combined_class = Cow::Borrowed("");
        if let Some(ref custom) = attrs.custom {
            for (key, value) in custom.iter() {
                if key == "style" {
                    if !combined_style.is_empty() {
                        combined_style.push_str("; ");
                    }
                    combined_style.push_str(value.as_str());
                } else if key == "class" {
                    combined_class = Cow::Borrowed(value.as_str());
                } else {
                    write!(self.out, " {}=\"{}\"", key, escape(value.as_str()))?;
                }
            }
        }

        if tag == Tag::FootnoteDefinition {
            combined_class = if combined_class.is_empty() {
                Cow::Borrowed(&self.options.footnote_definition_class)
            } else {
                Cow::Owned(format!(
                    "{} {}",
                    combined_class, &self.options.footnote_definition_class
                ))
            };
        }

        if !combined_class.is_empty() {
            write!(self.out, " class=\"{}\"", escape(&combined_class))?;
        }

        if !combined_style.is_empty() {
            write!(self.out, " style=\"{}\"", escape(&combined_style))?;
        }

        write!(self.out, ">")?;

        Ok(())
    }

    fn end_tag(&mut self, tag: Tag) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);

        write!(
            self.out,
            "</{}>{}",
            html_tag,
            if self.is_block_tag(tag) { "\n" } else { "" }
        )?;

        Ok(())
    }

    /// Feeds a single event into the renderer.
    pub fn feed_event(&mut self, event: &AnnotatedEvent<'data>) -> Result<(), io::Error> {
        match event.event {
            Event::DocumentStart(_) | Event::MetaData(_) => {}
            Event::StartTag(StartTagEvent { tag, ref attrs }) => {
                self.start_tag(tag, attrs)?;
            }
            Event::EndTag(EndTagEvent { tag }) => {
                self.end_tag(tag)?;
            }
            Event::Text(TextEvent { ref text }) => {
                write!(self.out, "{}", escape(text.as_str()))?;
            }
            Event::CodeBlock(CodeBlockEvent {
                ref code,
                ref language,
            }) => {
                write!(self.out, "<pre><code")?;
                if let Some(language) = language {
                    write!(self.out, " class=\"lang-{}\"", language.as_str())?;
                }
                writeln!(self.out, ">{}</code></pre>", escape(code.as_str()))?;
            }
            Event::Directive(DirectiveEvent {
                ref name, ref body, ..
            }) => {
                write!(
                    self.out,
                    "<div class=\"directive-{}\"><pre>{}</pre></div>",
                    escape(name.as_str()),
                    escape(body.as_str()),
                )?;
            }
            Event::InterpretedText(InterpretedTextEvent { ref text, ref role }) => {
                write!(
                    self.out,
                    "<span class=\"role-{}\">{}</span>",
                    escape(role.as_str()),
                    escape(text.as_str())
                )?;
            }
            Event::InlineCode(InlineCodeEvent { ref code }) => {
                write!(self.out, "<code>{}</code>", escape(code.as_str()))?;
            }
            Event::Image(ImageEvent {
                ref target,
                ref alt,
                ref title,
            }) => {
                write!(
                    self.out,
                    "<img src=\"{}\" alt=\"{}\" title=\"{}\">",
                    target,
                    alt.as_ref().map_or("", |x| x.as_str()),
                    title.as_ref().map_or("", |x| x.as_str()),
                )?;
            }
            Event::RawHtml(RawHtmlEvent { ref html }) => {
                write!(self.out, "{}", html)?;
            }
            Event::SoftBreak => writeln!(self.out)?,
            Event::HardBreak => writeln!(self.out, "<br>")?,
            Event::Rule => {
                write!(self.out, "<hr>")?;
            }
            Event::Checkbox(CheckboxEvent { checked }) => {
                write!(
                    self.out,
                    "<input type=checkbox disabled{}>",
                    if checked { " checked" } else { "" }
                )?;
            }
            Event::FootnoteReference(FootnoteReferenceEvent { ref target }) => {
                let number = match self.footnotes.get(target) {
                    Some(&num) => num,
                    None => {
                        let next_number = self.footnotes.len() + 1;
                        self.footnotes.insert(target.clone(), next_number);
                        next_number
                    }
                };
                write!(
                    self.out,
                    "<sup class=\"{}\"><a href=\"#{}\">{}</a></sup>",
                    escape(&self.options.footnote_reference_class),
                    escape(target.as_str()),
                    number,
                )?;
            }
            Event::Error(ErrorEvent {
                ref title,
                ref description,
            }) => {
                write!(
                    self.out,
                    "<div class=\"error\">\n<h3>{}</h3>\n<p>{}</p>\n</div>",
                    escape(title.as_str()),
                    escape(description.as_ref().map_or("No details", |x| x.as_str())),
                )?;
            }
        }
        Ok(())
    }

    /// Feeds an event stream into the renderer.
    pub fn feed_stream<I>(&mut self, iter: I) -> Result<(), io::Error>
    where
        I: Iterator<Item = AnnotatedEvent<'data>>,
    {
        for event in iter {
            self.feed_event(&event)?;
        }
        Ok(())
    }
}

impl<'data, 'options> HtmlRenderer<'data, 'options, Vec<u8>> {
    /// Creates a new html renderer writing into a buffer.
    pub fn new_buffered(options: &'options HtmlRendererOptions) -> Self {
        HtmlRenderer::new(Vec::new(), options)
    }

    /// Converts the renderer into a string.
    pub fn into_string(self) -> String {
        unsafe { String::from_utf8_unchecked(self.into_writer()) }
    }
}

/// Convenience shortcut that renders an event stream into HTML.
pub fn to_html<'a, I: Iterator<Item = AnnotatedEvent<'a>>>(
    iter: I,
    options: &HtmlRendererOptions,
) -> String {
    let mut renderer = HtmlRenderer::new_buffered(options);
    renderer.feed_stream(iter).unwrap();
    renderer.into_string()
}
