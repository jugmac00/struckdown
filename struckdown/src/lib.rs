//! struckdown is a text processing system based on markdown/commonmark with
//! structured extensions inspired by reStructuredText.  It parses commonmark
//! with structural extensions (roles and directives) into an event stream.
//!
//! This event stream can be processed before rendering.
pub mod event;
pub mod html;
pub mod parser;
pub mod pipeline;
pub mod processors;

// internal only for now
mod plain;

/// Gives access to [`serde_json`] value functionality.
///
/// The [`Value`](crate::value::Value) type is used to represent arbitrary data in a few instances.
/// A notable example is emitted [`front_matter`](crate::event::DocumentStartEvent::front_matter)
/// or other [`MetaData`](crate::event::MetaDataEvent).
pub mod value {
    pub use serde_json::value::*;

    /// A re-export of the [`serde_json::json`] macro.
    pub use serde_json::json as value;

    pub use serde_json::{from_value, to_value};
}
