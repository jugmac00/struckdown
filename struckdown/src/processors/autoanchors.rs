use slug::slugify;
use std::collections::VecDeque;

use crate::event::{AnnotatedEvent, Event, StartTagEvent, Tag};

/// Adds anchors to all headers if missing.
pub struct AutoAnchors<'data, I> {
    source: I,
    buffer: VecDeque<AnnotatedEvent<'data>>,
    min_level: usize,
}

impl<'data, I: Iterator<Item = AnnotatedEvent<'data>>> AutoAnchors<'data, I> {
    /// Creates a new auto anchor stream processor.
    ///
    /// This sets the default minimum level to 6 which means that all headers
    /// will get an ID added.  This minimum level can be reconfigured by using
    /// the [`AutoAnchors::set_min_level`] method.
    pub fn new(iterator: I) -> Self {
        Self {
            source: iterator,
            buffer: VecDeque::new(),
            min_level: 6,
        }
    }

    /// Overrides the minimum level that should get IDs
    pub fn set_min_level(&mut self, min_level: usize) {
        self.min_level = min_level;
    }
}

impl<'data, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator for AutoAnchors<'data, I> {
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(annotated_event) = self.buffer.pop_front() {
            return Some(annotated_event);
        }

        self.source.next().map(|mut annotated_event| {
            let (header_level, attrs) = match annotated_event.event {
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading1,
                    ref mut attrs,
                }) => (1, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading2,
                    ref mut attrs,
                }) => (2, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading3,
                    ref mut attrs,
                }) => (3, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading4,
                    ref mut attrs,
                }) => (4, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading5,
                    ref mut attrs,
                }) => (5, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading6,
                    ref mut attrs,
                }) => (6, attrs),
                _ => return annotated_event,
            };

            if attrs.id.is_some() || header_level > self.min_level {
                return annotated_event;
            }

            let mut depth = 1;
            let mut raw_text = String::new();

            while let Some(next_annotated_event) = self.source.next() {
                match next_annotated_event.event {
                    Event::StartTag(..) => depth += 1,
                    Event::EndTag(..) => depth -= 1,
                    ref event => {
                        if let Some(text) = event.raw_text() {
                            raw_text.push_str(text.as_str());
                        }
                    }
                }
                if depth == 0 {
                    break;
                }
            }

            attrs.id = Some(slugify(raw_text).into());

            annotated_event
        })
    }
}
