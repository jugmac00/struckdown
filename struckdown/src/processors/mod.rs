//! Stream processors for struckdown.
//!
//! This module provides stream processors that manipulate a struckdown event
//! stream to enhance it.  For instance a stream processor can automatically
//! add anchors to headers if they did not already get a header set by other
//! means.
mod autoanchors;

pub use self::autoanchors::{AutoAnchors, AutoAnchorsOptions};
