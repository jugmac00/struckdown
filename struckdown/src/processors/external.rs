use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::path::PathBuf;
use std::process::Stdio;
use tokio::runtime::Runtime;

use tokio::io::BufReader;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt};
use tokio::process::{ChildStdin, ChildStdout, Command};

use crate::event::AnnotatedEvent;
use crate::processors::Processor;

/// Passes a JSON serialized stream through an external program.
///
/// When applied this wraps the stream in a [`ExternalIter`].
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct External {
    /// The executable to run.
    pub cmd: PathBuf,
    /// The arguments to pass to the external command.
    #[serde(default)]
    pub args: Vec<String>,
}

impl Processor for External {
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(ExternalIter::new(iter, Cow::Owned(*self)))
    }

    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(ExternalIter::new(iter, Cow::Borrowed(self)))
    }
}

/// The iterator implementing [`External`].
pub struct ExternalIter<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> {
    source: I,
    spawned: bool,
    stdin: Option<ChildStdin>,
    stdout: Option<BufReader<ChildStdout>>,
    buffered_event: Option<Vec<u8>>,
    options: Cow<'options, External>,
    rt: Option<Runtime>,
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> ExternalIter<'data, 'options, I> {
    pub fn new(iterator: I, options: Cow<'options, External>) -> Self {
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .unwrap();
        Self {
            source: iterator,
            spawned: false,
            stdin: None,
            stdout: None,
            buffered_event: None,
            options,
            rt: Some(rt),
        }
    }
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator
    for ExternalIter<'data, 'options, I>
{
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        // TODO: error handling
        // This entire code is a massive hack.  There must be a better way to do this but
        // i'm too lazy to figure this out right now for a small experiment.

        let rt = self.rt.take().unwrap();
        if !self.spawned {
            rt.block_on(async {
                let mut process = Command::new(&self.options.cmd)
                    .args(&self.options.args)
                    .stdin(Stdio::piped())
                    .stdout(Stdio::piped())
                    .spawn()
                    .unwrap();
                self.stdin = process.stdin.take();
                self.stdout = process.stdout.take().map(BufReader::new);
                self.spawned = true;
            });
        }

        let mut done = false;
        while !done {
            if self.buffered_event.is_none() {
                if let Some(event) = self.source.next() {
                    let mut serialized_event = serde_json::to_vec(&event).unwrap();
                    serialized_event.push(b'\n');
                    self.buffered_event = Some(serialized_event);
                } else {
                    // close stdin if we're done writing.
                    self.stdin.take();
                }
            }

            let mut stdin = self.stdin.take();
            let mut stdout = self.stdout.take().unwrap();
            let mut rv = None;

            rt.block_on(async {
                let mut line = String::new();

                let should_write = self.buffered_event.is_some();
                let write = async {
                    if let (Some(ref mut stdin), Some(ref buffered_event)) = (&mut stdin, &self.buffered_event) {
                        stdin.write(buffered_event).await.unwrap();
                    }
                };

                tokio::select! {
                    read = stdout.read_line(&mut line) => {
                        match read {
                            Ok(0) | Err(_) => done = true,
                            Ok(_) => {
                                let event: AnnotatedEvent<'_> = serde_json::from_str(&line).unwrap();
                                rv = Some(event);
                            }
                        }
                    }
                    _ = write, if should_write => {
                        self.buffered_event.take();
                    }
                }
            });

            self.stdin = stdin;
            self.stdout = Some(stdout);

            if let Some(rv) = rv {
                self.rt = Some(rt);
                return Some(rv);
            } else if done {
                // close stuff
                self.stdin.take();
                self.stdout.take();
                self.rt = Some(rt);
                return None;
            }
        }

        None
    }
}
