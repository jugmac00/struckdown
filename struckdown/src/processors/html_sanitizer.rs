use std::borrow::Cow;
use std::collections::{BTreeMap, HashSet};

use ammonia::Builder;
use lazy_static::lazy_static;
use regex::Regex;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::event::{AnnotatedEvent, Event};
use crate::processors::Processor;

lazy_static! {
    static ref MARKER_RE: Regex = Regex::new(r"\A\.\.\.([a-f0-9]{32})\.\.\.").unwrap();
}

/// Automatically add anchors to all headers when missing.
///
/// When applied this wraps the stream in a [`HtmlSanitizerIter`].
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(default)]
pub struct HtmlSanitizer {
    /// Configures link rel
    link_rel: Option<String>,
    /// If configured the `class` attribute is permitted without filtering.
    allow_class: bool,
    /// If configured the `style` attribute is permitted without filtering.
    allow_style: bool,
    /// If set to `false` then comments are removed.
    allow_comments: bool,
}

impl Default for HtmlSanitizer {
    fn default() -> HtmlSanitizer {
        HtmlSanitizer {
            link_rel: Some("noopener noreferrer".into()),
            allow_class: false,
            allow_style: false,
            allow_comments: true,
        }
    }
}

impl Processor for HtmlSanitizer {
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(HtmlSanitizerIter::new(iter, Cow::Owned(*self)))
    }

    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(HtmlSanitizerIter::new(iter, Cow::Borrowed(self)))
    }
}

fn make_ammonia(options: &HtmlSanitizer) -> Builder {
    let mut ammonia = Builder::default();
    let mut clean_content_tags = HashSet::new();
    clean_content_tags.insert("script");

    if options.allow_class {
        ammonia.add_generic_attributes(&["class"]);
    }
    if options.allow_style {
        ammonia.add_generic_attributes(&["style"]);
        ammonia.add_tags(&["style"]);
    } else {
        clean_content_tags.insert("style");
    }
    ammonia.clean_content_tags(clean_content_tags);
    ammonia.link_rel(options.link_rel.as_deref());
    ammonia.strip_comments(!options.allow_comments);
    ammonia
}

/// The iterator implementing [`HtmlSanitizer`].
pub struct HtmlSanitizerIter<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> {
    source_iter: Option<I>,
    processed_iter: std::vec::IntoIter<AnnotatedEvent<'data>>,
    options: Cow<'options, HtmlSanitizer>,
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>>
    HtmlSanitizerIter<'data, 'options, I>
{
    pub fn new(iterator: I, options: Cow<'options, HtmlSanitizer>) -> Self {
        Self {
            source_iter: Some(iterator),
            processed_iter: Vec::new().into_iter(),
            options,
        }
    }
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator
    for HtmlSanitizerIter<'data, 'options, I>
{
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(source_iter) = self.source_iter.take() {
            let marker = format!("...{}...", Uuid::new_v4().to_simple());
            let mut buffer = source_iter.collect::<Vec<_>>();
            let mut html_buf = String::new();
            let ammonia = make_ammonia(&self.options);
            let mut segments = BTreeMap::new();
            let mut idx = 0;
            for annotated_event in &mut buffer {
                if let Event::RawHtml(ref mut raw_html) = annotated_event.event {
                    idx += 1;
                    let id = Uuid::new_v4();
                    html_buf.push_str(&format!("...{}...", id.to_simple()));
                    html_buf.push_str(raw_html.html.as_str());
                    html_buf.push_str(&marker);
                    segments.insert(id, idx);
                }
            }
            let cleaned = dbg!(ammonia.clean(&html_buf).to_string());

            let mut replacements = BTreeMap::new();
            for segment in cleaned.split(&marker) {
                if let Some(id) = MARKER_RE.captures(segment) {
                    if let Some(&idx) = segments.get(&Uuid::parse_str(&id[1]).unwrap()) {
                        replacements.insert(idx, &segment[id.get(0).unwrap().end()..]);
                    }
                }
            }

            let mut idx = 0;
            for annotated_event in &mut buffer {
                if let Event::RawHtml(ref mut raw_html) = annotated_event.event {
                    idx += 1;
                    if let Some(&new) = replacements.get(&idx) {
                        raw_html.html = new.to_string().into();
                        continue;
                    }
                    raw_html.html = "".into();
                }
            }

            self.processed_iter = buffer.into_iter();
        }
        self.processed_iter.next()
    }
}
