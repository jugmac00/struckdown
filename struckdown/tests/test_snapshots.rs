use std::fs;

use struckdown::event::{AnnotatedEvent, DocumentStartEvent, Event};
use struckdown::html::to_html;
use struckdown::parser::parse;
use struckdown::processors;

use itertools::Either;

use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(tag = "processor")]
pub enum Processor {
    AutoAnchors(processors::AutoAnchorsOptions),
}

fn apply_configured_processors<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    processors: Vec<serde_yaml::Value>,
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut iter = Box::new(iter) as Box<dyn Iterator<Item = AnnotatedEvent<'data>>>;

    for processor in processors {
        let processor: Processor = serde_yaml::from_value(processor).unwrap();
        match processor {
            Processor::AutoAnchors(options) => {
                iter = Box::new(processors::AutoAnchors::new_with_options(iter, options));
            }
        }
    }

    return iter;
}

fn apply_processors<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut iter = iter.peekable();

    if let Some(document_start) = iter.peek() {
        match document_start.event {
            Event::DocumentStart(DocumentStartEvent {
                front_matter: Some(ref front_matter),
            }) => {
                if let Some(processors) =
                    front_matter.get("processors").and_then(|x| x.as_sequence())
                {
                    return Either::Left(apply_configured_processors(
                        processors.iter().map(|x| x.clone()).collect(),
                        iter,
                    ));
                }
            }
            _ => {}
        }
    }

    Either::Right(iter)
}

#[test]
fn test_parser() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let events: Vec<_> = apply_processors(parse(&source)).collect();
        insta::assert_yaml_snapshot!(events);
    });
}

#[test]
fn test_html() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let html = to_html(apply_processors(parse(&source)));
        insta::assert_snapshot!(html);
    });
}
