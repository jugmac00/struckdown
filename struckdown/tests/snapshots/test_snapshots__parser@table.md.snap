---
source: struckdown/tests/test_snapshots.rs
expression: events
input_file: struckdown/tests/inputs/table.md
---
- type: document_start
- - type: start_tag
    tag: paragraph
  - offset: 0
    len: 28
    line: 1
    column: 0
- - type: text
    text: "| a | b | c |"
  - offset: 0
    len: 13
    line: 1
    column: 0
- - type: soft_break
  - offset: 13
    len: 1
    line: 1
    column: 13
- - type: text
    text: "| 1 | 2 | 3 |"
  - offset: 14
    len: 13
    line: 2
    column: 0
- - type: end_tag
    tag: paragraph
  - offset: 0
    len: 28
    line: 1
    column: 0
- - type: start_tag
    tag: table
  - offset: 29
    len: 60
    line: 4
    column: 0
- type: start_tag
  tag: table_header
- - type: start_tag
    tag: table_head
  - offset: 31
    len: 1
    line: 4
    column: 2
- - type: text
    text: a
  - offset: 31
    len: 1
    line: 4
    column: 2
- - type: end_tag
    tag: table_head
  - offset: 31
    len: 1
    line: 4
    column: 2
- - type: start_tag
    tag: table_head
  - offset: 37
    len: 1
    line: 4
    column: 8
- - type: text
    text: b
  - offset: 37
    len: 1
    line: 4
    column: 8
- - type: end_tag
    tag: table_head
  - offset: 37
    len: 1
    line: 4
    column: 8
- - type: start_tag
    tag: table_head
  - offset: 43
    len: 1
    line: 4
    column: 14
- - type: text
    text: c
  - offset: 43
    len: 1
    line: 4
    column: 14
- - type: end_tag
    tag: table_head
  - offset: 43
    len: 1
    line: 4
    column: 14
- - type: end_tag
    tag: table_header
  - offset: 29
    len: 20
    line: 4
    column: 0
- type: start_tag
  tag: table_body
- - type: start_tag
    tag: table_row
  - offset: 69
    len: 20
    line: 6
    column: 0
- - type: start_tag
    tag: table_cell
  - offset: 71
    len: 1
    line: 6
    column: 2
- - type: text
    text: "1"
  - offset: 71
    len: 1
    line: 6
    column: 2
- - type: end_tag
    tag: table_cell
  - offset: 71
    len: 1
    line: 6
    column: 2
- - type: start_tag
    tag: table_cell
  - offset: 77
    len: 1
    line: 6
    column: 8
- - type: text
    text: "2"
  - offset: 77
    len: 1
    line: 6
    column: 8
- - type: end_tag
    tag: table_cell
  - offset: 77
    len: 1
    line: 6
    column: 8
- - type: start_tag
    tag: table_cell
  - offset: 83
    len: 1
    line: 6
    column: 14
- - type: text
    text: "3"
  - offset: 83
    len: 1
    line: 6
    column: 14
- - type: end_tag
    tag: table_cell
  - offset: 83
    len: 1
    line: 6
    column: 14
- - type: end_tag
    tag: table_row
  - offset: 69
    len: 20
    line: 6
    column: 0
- type: end_tag
  tag: table_body
- - type: end_tag
    tag: table
  - offset: 29
    len: 60
    line: 4
    column: 0
- - type: start_tag
    tag: table
  - offset: 90
    len: 60
    line: 8
    column: 0
- type: start_tag
  tag: table_header
- - type: start_tag
    tag: table_head
    attrs:
      alignment: left
  - offset: 92
    len: 1
    line: 8
    column: 2
- - type: text
    text: a
  - offset: 92
    len: 1
    line: 8
    column: 2
- - type: end_tag
    tag: table_head
  - offset: 92
    len: 1
    line: 8
    column: 2
- - type: start_tag
    tag: table_head
    attrs:
      alignment: center
  - offset: 99
    len: 1
    line: 8
    column: 9
- - type: text
    text: b
  - offset: 99
    len: 1
    line: 8
    column: 9
- - type: end_tag
    tag: table_head
  - offset: 99
    len: 1
    line: 8
    column: 9
- - type: start_tag
    tag: table_head
    attrs:
      alignment: right
  - offset: 106
    len: 1
    line: 8
    column: 16
- - type: text
    text: c
  - offset: 106
    len: 1
    line: 8
    column: 16
- - type: end_tag
    tag: table_head
  - offset: 106
    len: 1
    line: 8
    column: 16
- - type: end_tag
    tag: table_header
  - offset: 90
    len: 20
    line: 8
    column: 0
- type: start_tag
  tag: table_body
- - type: start_tag
    tag: table_row
  - offset: 130
    len: 20
    line: 10
    column: 0
- - type: start_tag
    tag: table_cell
    attrs:
      alignment: left
  - offset: 132
    len: 1
    line: 10
    column: 2
- - type: text
    text: "1"
  - offset: 132
    len: 1
    line: 10
    column: 2
- - type: end_tag
    tag: table_cell
  - offset: 132
    len: 1
    line: 10
    column: 2
- - type: start_tag
    tag: table_cell
    attrs:
      alignment: center
  - offset: 139
    len: 1
    line: 10
    column: 9
- - type: text
    text: "2"
  - offset: 139
    len: 1
    line: 10
    column: 9
- - type: end_tag
    tag: table_cell
  - offset: 139
    len: 1
    line: 10
    column: 9
- - type: start_tag
    tag: table_cell
    attrs:
      alignment: right
  - offset: 146
    len: 1
    line: 10
    column: 16
- - type: text
    text: "3"
  - offset: 146
    len: 1
    line: 10
    column: 16
- - type: end_tag
    tag: table_cell
  - offset: 146
    len: 1
    line: 10
    column: 16
- - type: end_tag
    tag: table_row
  - offset: 130
    len: 20
    line: 10
    column: 0
- type: end_tag
  tag: table_body
- - type: end_tag
    tag: table
  - offset: 90
    len: 60
    line: 8
    column: 0
